### PR TITLE
Implementation of recursive vmla transpose kernel with test.

### PR DESCRIPTION
**Pull request by KoolJBlack**
Very fast recursive implementation of transpose for VMLA. Does the minimum number of computations necessary for transpose: computes and caches intermediate offsets while going down the stack and copies each input value directly to its final output position.

Mean Time Per Call (MTPC) of TransposeX32 for bert splats went from 1.95ms to 136us

Before:
transpose_before_optimization

After:
transpose_after_optimization

Contributes to #3601

### CODE REVIEW

**Comment by benvanik** [comment]
Nice! If tile and reduce get the same treatment we may actually match/beat tflite :P

File: iree/hal/vmla/op_kernels_generic.h [hunk 1]
```diff
@@ -238,32 +238,62 @@ Status Finite::Execute(absl::Span<const T> src_buffer,
...
+    for (size_t i = 0; i < dst_shape[dim_i]; ++i) {
+      size_t src_offset = src_base_offset + i * src_strides[perm[dim_i]];
+      size_t dst_offset = dst_base_offset + i * dst_strides[dim_i];
+      TransposeRecurse(src_buffer, dst_buffer, src_shape, dst_shape,
+                       src_strides, dst_strides, perm, rank, recurse_dim_i,
+                       src_offset, dst_offset);
+    }
+  } else {
+    for (size_t i = 0; i < dst_shape[dim_i]; ++i) {
+      size_t src_i = src_base_offset + i * src_strides[perm[dim_i]];
```
Comment (benvanik) [inline iree/hal/vmla/op_kernels_generic.h:263]: the compiler may be doing this for you but you could try moving the stride lookup outside of the loop here (and above) as it's constant per dim_i

**Review by benvanik — APPROVED** [review]

File: iree/hal/vmla/op_kernels_test.cc [hunk 2]
```diff
@@ -536,6 +536,49 @@ TEST(Conv2d, DepthwiseConv) {
...
+TEST(Transpose, 2Dimen) {
+  Shape src_shape = {2, 3};
+  Shape dst_shape = {3, 2};
+  std::vector<int32_t> perm = {1, 0};
+  // clang-format off
+  std::vector<uint16_t> src_buffer = { 1,  2,   3,
+                                      4,  5,   6};
+  std::vector<uint16_t> expected_dst = { 1,  4,
+                                         2,  5,   
+                                         3,  6};
```
Comment (ScottTodd) [inline iree/hal/vmla/op_kernels_test.cc:548]: nit: if you disable clang-format for some lines, at least make them pretty ;D
```suggestion
  std::vector<uint16_t> src_buffer = {1, 2, 3,
                                      4, 5, 6};
  std::vector<uint16_t> expected_dst = {1, 4,
                                        2, 5,
                                        3, 6};
```

You could also work with clang-format by inserting comments to block its standard line wrapping:

```c++
  // (no clang-format off)
  std::vector<uint16_t> src_buffer = {1, 2, 3,  //
                                      4, 5, 6};
  std::vector<uint16_t> expected_dst = {1, 4,  //
                                        2, 5,  //
                                        3, 6};
```

File: iree/hal/vmla/op_kernels_test.cc [hunk 3]
```diff
@@ -536,6 +536,49 @@ TEST(Conv2d, DepthwiseConv) {
...
+                                      4,  5,  6,
+                                      7,  8,  9,
+                                      10, 11, 12};
+  std::vector<uint16_t> expected_dst = { 1,  4,
+                                         2,  5,   
+                                         3,  6,
+                                         7,  10,
+                                         8,  11,
+                                         9,  12};
+  // clang-format on
```
Comment (ScottTodd) [inline iree/hal/vmla/op_kernels_test.cc:573]: Same down here. At least remove the trailing spaces after `2,  5,`

```suggestion
  // clang-format off
  std::vector<uint16_t> src_buffer = { 1,  2,  3,
                                       4,  5,  6,
                                       7,  8,  9,
                                      10, 11, 12};
  std::vector<uint16_t> expected_dst = {1,  4,
                                        2,  5,
                                        3,  6,
                                        7, 10,
                                        8, 11,
                                        9, 12};
  // clang-format on
```

**Review by ScottTodd — APPROVED** [review]
👍 nice, thanks for the Tracy screenshots!

File: iree/hal/vmla/op_kernels_test.cc (hunk 2 above)
Comment (KoolJBlack) [reply iree/hal/vmla/op_kernels_test.cc:548]: I've updated the indexing and also surrounded the values in block quotes to disable clang format.

**Pull request merged**